WebProcess hung on main thread in WebPage::platformInitializeAccessibility

pvollan opened the pull request:
#### 5bcf0b7db2735a8090596c182615150e5d1c9f03
<pre>
WebProcess hung on main thread in WebPage::platformInitializeAccessibility
<a href="https://bugs.webkit.org/show_bug.cgi?id=258140">https://bugs.webkit.org/show_bug.cgi?id=258140</a>
rdar://110191499

Reviewed by Brent Fulgham.

The WebProcess can hang on the main thread in WebPage::platformInitializeAccessibility for several seconds.
This appears to be caused by the call to [NSApplication _accessibilityInitialize] trying to connect to the
Launch Services daemon. At this point, the daemon is blocked in the sandbox. Fix this by extending the time
holding the sandbox extension to the daemon, which has been revoked.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::platformInitializeAccessibility):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformInitializeWebProcess):
(WebKit::WebProcess::revokeLaunchServicesSandboxExtension):
* Source/WebKit/WebProcess/cocoa/LaunchServicesDatabaseManager.h:
(WebKit::LaunchServicesDatabaseManager::hasReceivedLaunchServicesDatabase const):

Canonical link: <a href="https://commits.webkit.org/265334@main">https://commits.webkit.org/265334@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b2794c1b36cf2e0734ec5de57b66bc7619f764b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10654 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10879 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11163 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12303 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10224 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13249 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10848 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13142 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10815 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11732 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8957 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12706 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9026 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9616 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/9018 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10096 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/9767 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13024 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/10110 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10236 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/10772 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9390 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/2904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2544 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13664 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/11072 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10097 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2710 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
```diff
@@ -131,6 +137,13 @@
     accessibilityTransferRemoteToken(accessibilityRemoteTokenData());
+
+    // Close Mach connection to Launch Services.
+    // FIXME: Replace the constant 4 with kLSServerConnectionStatusReleaseNotificationsMask when available in the SDK.
```
I think it would be tidier to have a define for `kLSServerConnectionStatusReleaseNotificationsMask` protected by something like `#if !HAVE(LS_DONOTCONNECTMASK)`. That way the code using the definition never has to change, just a future removal of the local define.

pvollan replied:
Fixed!

Thanks for reviewing!

brentfulgham commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -404,25 +406,10 @@ static void softlinkDataDetectorsFrameworks()
-    
-    // This is being called under WebPage::platformInitialize(), and may reach out to the Launch Services daemon once in the lifetime of the process.
-    // Call this synchronously here while a sandbox extension to Launch Services is being held.
-    [NSAccessibilityRemoteUIElement remoteTokenForLocalUIElement:adoptNS([[WKAccessibilityWebPageObject alloc] init]).get()];
```
How is it safe to remove this bit?

pvollan replied:
That is a very good point. This is actually covered in the call to accessibilityRemoteTokenData() in WebPage::platformInitializeAccessibility. This bit was copied there to work around a race with closing the Mach connection to Launch Services.

brentfulgham reviewed (approved):
Looks great! Thank you.

pvollan committed:
WebProcess hung on main thread in WebPage::platformInitializeAccessibility
https://bugs.webkit.org/show_bug.cgi?id=258140
rdar://110191499

Reviewed by Brent Fulgham.

The WebProcess can hang on the main thread in WebPage::platformInitializeAccessibility for several seconds.
This appears to be caused by the call to [NSApplication _accessibilityInitialize] trying to connect to the
Launch Services daemon. At this point, the daemon is blocked in the sandbox. Fix this by extending the time
holding the sandbox extension to the daemon, which has been revoked.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::platformInitializeAccessibility):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformInitializeWebProcess):
(WebKit::WebProcess::revokeLaunchServicesSandboxExtension):
* Source/WebKit/WebProcess/cocoa/LaunchServicesDatabaseManager.h:
(WebKit::LaunchServicesDatabaseManager::hasReceivedLaunchServicesDatabase const):

Canonical link: https://commits.webkit.org/265334@main
